packages/devtools_app/lib/src/screens/debugger/breakpoints.dart

```diff
@@ -19,28 +17,29 @@ double get executableLineRadius => scaleByFontFactor(1.5);
 double get breakpointRadius => scaleByFontFactor(6.0);
 
 class Breakpoints extends StatefulWidget {
-  const Breakpoints({Key key}) : super(key: key);
+  const Breakpoints({Key? key}) : super(key: key);
 
   @override
   _BreakpointsState createState() => _BreakpointsState();
 }
 
 class _BreakpointsState extends State<Breakpoints> {
-  DebuggerController controller;
+  DebuggerController get controller => _controller!;
+  DebuggerController? _controller;
```

Comment:
Another pattern we've used elsewhere in DevTools is making the controller var late and using a bool to track whether or not we have initialized the controller. This avoids having to perform a null-assert every time the controller getter is accessed. See https://github.com/flutter/devtools/blob/3f83c3597740677a5d86b7808e2ac5c8addba409/packages/devtools_app/lib/src/screens/performance/performance_screen.dart#L101-L103

Reply:
> Another pattern we've used elsewhere in DevTools is making the controller var late and using a bool to track whether or not we have initialized the controller. This avoids having to perform a null-assert every time the controller getter is accessed. See
> 
> https://github.com/flutter/devtools/blob/3f83c3597740677a5d86b7808e2ac5c8addba409/packages/devtools_app/lib/src/screens/performance/performance_screen.dart#L101-L103

What is the motivation for using this pattern rather than defining the getter + setter pair?
There isn't much magic to `late`. It basically just desugars to the getter setter pair. Keeping `_controllerInitialized` separate from the field leaves room for bugs if one is set but the other isn't and uses more memory as an extra boolean field is tracked.
With late, the Dart runtime still has to do the `!` check, you just didn't have to explicitly write it. There are a few cases where the AOT compiler is clever about optimizing the bang checks for late in ways that it may not for the desugared version but in general that shouldn't matter for DevTools.

Reply:
> With late, the Dart runtime still has to do the ! check, you just didn't have to explicitly write it. 

I was unaware of this. If calling `controller!` is the same thing that would be happening under the hood for `late controller`, then there is no benefit to using late. CC @bkonyi who originally introduced this pattern in case you had any motivations behind using late over a getter setter pair that I may be missing.

Reply:
Yeah there is no free lunch. `late` still requires the runtime to verify the member is non-null before use.
Another valid case to take `late` variables and switch them to a getter + setter pair is if you need to null the variable in the `dispose()` method.

Reply:
Yeah, using `late` introduces an unavoidable sentinel check since we need to be able to throw a late initialization error if the variable isn't initialized. I still prefer this pattern over making `_controller` nullable if we never need to explicitly set the controller to `null` since the null assertions give the impression that the check _could_ fail, even if it never will.

Reply:
`late` should give you the same impression that any accesses to the late variable could fail with a null assertion. There is no compile time checking that a late variable is initialized properly before use. `late` is just as much of a runtime feature as the getter+field pair. It is 95% just syntactic sugar so you don't have to manually write the getter + field pair.

Reply:
Switched to the non-null getter / nullable private field pattern.